### PR TITLE
Allow custom coffeelint-reporters

### PR DESCRIPTION
**Pull request by jayharris**
The current `gulp-coffeelint` implementation only supports the plugin's `default`, `fail`, and `failOnWarning` reporters. My intention was to extend the Reporter system to be more like `gulp-jshint`, allowing developers to specify other Reporters, including CoffeeLint's built-in reporters, module reporters, or even custom functions.

This PR aims to maintain the current functionality, `coffeelint.reporter()` and `coffeelint.reporter('default')` will both still invoke `coffeelint-stylish`, but now passing a string, function, or coffeelint-reporter to the function will now use that other reporter.

Much of the idea was poached from `gulp-jshint`. The documentation has also been updated, also primarily poached from `gulp-jshint`.

I tried to keep all styling the same. All tests pass, including new unit tests.

One note regarding maintaining functionality: CoffeeLint includes a `default` build-in reporter. By maintaining existing functionality, `coffeelint.reporter('default')` will invoke `coffeelint-stylish` rather than CoffeeLint's built-in default reporter. This is a difference in functionality from JSHint.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -64,28 +64,83 @@ file.coffeelint.literate = false; // you guessed it
...
+  .pipe(coffeelint.reporter('coffeelint-stylish'))
   .pipe(coffeelint.reporter('fail'))
 ```
 
+### Custom Reporters
+
+Custom reporters don't interact with this module at all. CoffeeLint will add some attributes to the file object and you can add a custom reporter downstream.
+
+```js
+var jshint = require('gulp-coffeelint');
```
Comment (janraasch) [inline readme.md]: c&p :)

File: lib/reporter.coffee [hunk 2]
```diff
@@ -44,12 +29,43 @@ failOnWarningReporter = ->
...
 
+loadReporter = (reporter) ->
+    return reporter if typeof reporter is 'function'
+    if typeof reporter is 'object' and typeof reporter.reporter is 'function'
+        return reporter.reporter
+
+    if typeof reporter is 'string'
+        # Try to load CoffeeLint's build-in reporters
+        try
+            return loadReporter require('coffeelint/src/reporters/' + reporter)
```
Comment (janraasch) [inline lib/reporter.coffee]: Requiring `.coffee` files would make the `coffee-script` npm module a production dependency, which is really bad style for a npm module.

**Comment by jayharris** [comment]
Makes complete sense. I want to rethink the implementation on this, anyway. I'll update and resubmit. I also missed a few test conditions that I would like to cover.

One question for you, @janraasch: Do you think it is worthwhile changing the explicit string `default` to mean the default CoffeeLint reporter instead of `coffeelint-stylish`? I think there's value in allowing consumers to use CoffeeLint's default reporter, but this would be a breaking change to anyone actually using `coffeelint.reporter('default')`.

In other words: currently `coffeelint.reporter()` and `coffeelint.reporter('default')` both trigger `coffeelint-stylish`. I think it makes sense to keep `coffeelint.reporter()` as `coffeelint-stylish`, but I think `coffeelint.reporter('default')` be switched to use the default CoffeeLint reporter, instead. Otherwise, consumers may say "I specified `default`, so why am I still getting `coffeelint-stylish`?"

What do you think?

**Comment by janraasch** [comment]
Sure, we could do both. Have one release introducing the new custom reporters (keeping the existing API), see how it works in the wild and have a major release with saner defaults in couple of weeks. 

Could you post a screenshot of the output of the *real* default `coffeelint` reporter used in combination with this `gulp` task? Say e.g. have **two** (to make the output interesting: both should have errors or warnings.) files linted by `gulp-coffelint. I would like to have reference screenshot to see, if that is a good default choice.

**Comment by jayharris** [comment]
Canceling this pull request. Going to cleanup my code a bit and submit a new PR.
